shiny/ui/_chat.py

```diff
@@ -738,34 +740,57 @@ def _store_message(
 
         return msg
 
-    def _get_trimmed_messages(
-        self,
-        *,
-        token_limits: tuple[int, int] | None = (4096, 1000),
+    @staticmethod
+    def _trim_messages(
+        messages: tuple[StoredMessage, ...],
+        token_limits: tuple[int, int] = (4096, 1000),
     ) -> tuple[StoredMessage, ...]:
-        messages = self._messages()
-
-        if token_limits is None:
-            return messages
-
-        # Can't trim if we don't have token counts
-        token_counts = [m["token_count"] for m in messages]
-        if None in token_counts:
-            return messages
 
-        token_counts = cast("list[int]", token_counts)
+        system_token_count: int = 0
+        token_counts: list[int] = []
+        system_messages: list[str] = []
+        other_messages: list[str] = []
+        for m in messages:
+            count = m["token_count"]
+            # Count can be None if the tokenizer is None
+            if count is None:
+                return messages
+            content = m["content_server"]
+            if m["role"] == "system":
+                system_messages.append(content)
+                system_token_count += count
+            else:
+                other_messages.append(content)
+                token_counts.append(count)
 
-        # Take the newest messages up to the token limit
         limit, reserve = token_limits
         max_tokens = limit - reserve
+
+        if system_token_count >= max_tokens:
+            raise ValueError(
+                f"System messages exceed `.messages(token_limits={token_limits})`. "
+                "Consider increasing the 1st value of `token_limit` or setting it to "
+                "`token_limit=None` to disable token limits."
+            )
+
+        max_tokens -= system_token_count
+
         messages2: list[StoredMessage] = []
         for i, m in enumerate(reversed(messages)):
-            if sum(token_counts[-i - 1 :]) > max_tokens:
-                break
-            messages2.append(m)
+            if m["role"] == "system":
+                messages2.append(m)
+            elif sum(token_counts[-i - 1 :]) <= max_tokens:
```

Comment:
I think that `token_counts[-i - 1 :]` will result in incorrect indexing in some edge cases. The `messages` data structure could have a system message that is not the first message. In the code above, where it does the first pass iterating over `messages`, it appends to `token_counts` only for non-system messages.

However, in this loop where it iterates over `messages`, it indexes into `token_counts` using `i`, where the `i` also is incremented with system messages -- but the `token_counts` list doesn't include counts for system messages.

To fix this, I think it would be simpler to replace `max_tokens` with `remaining_non_system_tokens`, and subtract from it in each iteration of the loop.

Reply:
Thanks! Addressed in [215109d](https://github.com/posit-dev/py-shiny/pull/1509/commits/215109d8f4743f646f1a1e65e5a75cc7f049758a)